router/dataplane_spec.gobra

```diff
@@ -343,3 +349,262 @@ func closureSpec2(i uint16, c BatchConn)
+}
+
+//expects the raw bytes of whole path
+ghost 
+requires 4 + 8 * numINF + offset*12 + 12 <= len(raw)
+requires prevSegLen >= 0 && offset - prevSegLen >= -1
+requires 0 < numINF && numINF < 4
+requires acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), _)
+decreases offset - prevSegLen + 1
+pure func getSegHistory(raw []byte, prevSegLen int, offset int, numINF int) seq[io.IO_ahi]{
+	return offset - prevSegLen == -1 ? seq[io.IO_ahi]{} : 
+		(unfolding acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), _) in 
+			seq[io.IO_ahi]{getHopField(raw,4 + 8 * numINF + offset*12).Toab()}) ++ getSegHistory(raw, prevSegLen, offset-1, numINF)
+}
+
+
+
+ghost 
+//expects the raw bytes of whole path
+requires 4 + 8 * numINF + 12*prevSegLen + 12*segLen <= len(raw)
+requires prevSegLen >= 0 && segLen >= 0
+requires 0 < numINF && numINF < 4
+requires segLen+prevSegLen > currHF && currHF >= prevSegLen
+requires acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), _)
+decreases 
+pure func getSegment(raw []byte, consDir bool, peer bool, currHF int, prevSegLen int, segLen int, numINF int, currINF int) io.IO_seg2{
+	return io.IO_seg2(io.IO_seg3_{
+				AInfo : computeAInfo(raw, currINF),
+				UInfo : computeUInfo(raw, currINF),
+				ConsDir : consDir,
+				Peer : peer,
+				Past : getSegPast(raw, prevSegLen, currHF - 1, numINF),
+				Future : getSegFuture(raw, currHF, segLen+prevSegLen-1, numINF),
+				History : getSegHistory(raw, prevSegLen, currHF - 1, numINF),
+			})
+}
+
+ghost 
+//expects the raw bytes of whole path
+requires 4 + 8 * numINF + 12*prevSegLen + 12*segLen <= len(raw)
+requires prevSegLen >= 0 && segLen >= 0
+requires 0 < numINF && numINF < 4
+requires segLen+prevSegLen > currHF && currHF >= prevSegLen
+requires acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), _)
+decreases 
+pure func getPastSegment(raw []byte, consDir bool, peer bool, currHF int, prevSegLen int, segLen int, numINF int, currINF int) io.IO_seg2{
+	return io.IO_seg2(io.IO_seg3_{
+				AInfo : computeAInfo(raw, currINF),
+				UInfo : computeUInfo(raw, currINF),
+				ConsDir : consDir,
+				Peer : peer,
+				Past : getSegPast(raw, prevSegLen, currHF, numINF),
+				Future : seq[io.IO_HF]{},
+				History : getSegHistory(raw, prevSegLen, currHF, numINF),
+			})
+}
+
+
+ghost
+requires seg1Len > 0
+requires seg2Len >= 0
+requires seg3Len >= 0
+requires 4 + 8*numInfoFields(seg1Len, seg2Len, seg3Len) + 12 * seg1Len + 12 * seg2Len + 12 * seg3Len == len(raw)
+requires 0 <= currINF && currINF < 3
+requires 0 <= currHF && currHF < seg1Len + seg2Len + seg3Len
+requires computeSeqlen(currHF, seg1Len, seg2Len, seg3Len) + computeprevSegLen(currHF, seg1Len, seg2Len, seg3Len) > currHF
+requires acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), _)
+decreases 
+pure func getCurrSeg(raw []byte, currINF int, currHF int, seg1Len int, seg2Len int, seg3Len int) io.IO_seg3 {
+	return let consDir := (unfolding acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), _) in raw[infoFieldOffsetSeg(currINF)] & 0x1 == 0x1) in
```

Comment:
can you introduce an intermediate pure func to avoid having to write all these unfoldings explicitly?